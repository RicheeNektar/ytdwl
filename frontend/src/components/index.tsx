export { default as RoundedWrapper } from './rounded-wrapper';
export { default as ProgressBar } from './progress-bar';
export { default as Thumbnail } from './thumbnail';
export { default as Video } from './video';
