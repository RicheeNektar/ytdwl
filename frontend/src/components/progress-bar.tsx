import styled from 'styled-components';

type Props = {
  className?: string;
  isAudio: boolean;
  progress: number;
  total: number;
};

const defaultProps: Pick<Props, 'isAudio'> = {
  isAudio: true,
};

const Bar = styled.div<Props>`
  height: 100%;
  display: block;
  position: relative;
  border-radius: 10px;
  margin-bottom: 5px;
  transition: ease-out 0.1s;
  box-shadow: inset 0 0 4px ${p => p.theme.gray100};
  background: ${p => (p.isAudio ? p.theme.audio : p.theme.video)};
  width: ${p => (p.progress / p.total) * 100}%;
`;

const Wrapper = styled.div`
  width: 100%;
  height: 0.5rem;
`;

const ProgressBar = ({ progress, total, isAudio, className }: Props) => {
  const renderBar = () => (
    <Bar progress={progress} total={total} isAudio={isAudio} />
  );

  return <Wrapper className={className}>{renderBar()}</Wrapper>;
};

ProgressBar.defaultProps = defaultProps;
export default ProgressBar;
