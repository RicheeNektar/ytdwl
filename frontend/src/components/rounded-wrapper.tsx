import styled from 'styled-components';

const Wrapper = styled.div`
  & > :first-child {
    border-radius: 10px 10px 0 0;
  }
  & > :last-child {
    border-radius: 0 0 10px 10px;
  }
`;

type Props = {
  children?: JSX.Element | JSX.Element[];
};

const RoundedWrapper = ({ children }: Props) => <Wrapper>{children}</Wrapper>;

export default RoundedWrapper;
