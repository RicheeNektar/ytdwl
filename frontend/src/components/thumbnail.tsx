import styled from 'styled-components';

const StyledPicture = styled.picture`
  max-height: 64px;
`;

const StyledImg = styled.img`
  border-radius: 10px;
  max-height: inherit;
  height: 100%;
  width: auto;
  aspect-ratio: 16/9;
`;

type Props = {
  videoId: string;
  className?: string;
};

const Thumbnail = ({ className, videoId }: Props) => {
  const src = (size: string) => `https://i.ytimg.com/vi/${videoId}/${size}.jpg`;

  const renderSource = (size: string) => (
    <source key={size} srcSet={src(size)} />
  );

  return (
    <StyledPicture className={className}>
      {['default', 'hqdefault'].map(renderSource)}
      <StyledImg
        key="maxresdefault"
        srcSet={src('maxresdefault')}
        loading="lazy"
      />
    </StyledPicture>
  );
};

export default Thumbnail;
