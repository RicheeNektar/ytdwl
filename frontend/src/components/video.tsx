import { ProgressBar, Thumbnail } from 'components';
import styled from 'styled-components';

const Wrapper = styled.div<{ tabId?: number }>`
  ${p => p.tabId && 'cursor: pointer;'}
  padding: 0.5rem;
  user-select: none;
  background-color: #555;
  box-shadow: 2px 2px 3px #333;
`;

const ContentWrapper = styled.div`
  display: flex;
  width: 100%;
  font-size: 10px;
`;

const VideoInfoWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  width: 100%;
  margin-left: 0.5rem;
`;

const StyledProgressBar = styled(ProgressBar)`
  margin-top: 0.25rem;
`;

type Props = {
  progressTotal?: number;
  showProgress?: boolean;
  progress?: number;
  isAudio?: boolean;
  tabId?: number;
  video: Video;
};

const defaultProps: Pick<Props, 'isAudio' | 'showProgress'> = {
  showProgress: false,
  isAudio: true,
};

const Video = ({
  progressTotal,
  showProgress,
  progress,
  isAudio,
  tabId,
  video: { title, id },
}: Props) => (
  <Wrapper
    tabId={tabId}
    key={id}
    id={id}
    {...(tabId && { title: 'Switch to tab' })}
  >
    <ContentWrapper>
      <Thumbnail videoId={id} />

      <VideoInfoWrapper>
        <span>{title}</span>
        <span>{isAudio ? 'Audio (.weba)' : 'Video (.webm)'}</span>
      </VideoInfoWrapper>
    </ContentWrapper>

    {showProgress && progressTotal && progress && (
      <StyledProgressBar
        isAudio={isAudio}
        progress={progress}
        total={progressTotal}
      />
    )}
  </Wrapper>
);

Video.defaultProps = defaultProps;
export default Video;
