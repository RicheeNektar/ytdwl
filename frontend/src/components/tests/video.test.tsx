import TestRenderer from 'react-test-renderer';
import { Video as VideoComponent } from 'components';
import { ComponentProps } from 'react';

const video = {
  id: 'test-id',
  title: 'Lorem Ipsum LIVE 2021',
};

type Props = Omit<ComponentProps<typeof VideoComponent>, 'video'>;

const createComponent = (props?: Props) =>
  TestRenderer.create(<VideoComponent video={video} {...props} />);

const standardTest = (props?: Props) => {
  const component = createComponent(props);
  const title = component.root.findAllByType('span')[0];

  expect(title.children.some(e => e === video.title)).toBeTruthy();
  expect(
    component.root.findAll(e => e.props?.srcSet?.includes(video.id)).length
  ).toBe(4);

  return component;
};

describe('Video', () => {
  it('renders without error', () => {
    const component = standardTest();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with switch to tab', () => {
    const component = standardTest({ tabId: 1 });

    expect(
      component.root.findAllByType('div')[0].props.title
    ).not.toBeUndefined();

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with progress bar', () => {
    const component = standardTest({
      showProgress: true,
      progressTotal: 2,
      progress: 1,
    });

    expect(component.root.findAllByType('div')[0].children.length).toBe(2);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
