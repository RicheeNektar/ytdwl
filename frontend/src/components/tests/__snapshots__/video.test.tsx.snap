// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video renders with progress bar 1`] = `
<div
  className="sc-jSFjdj kNXPLX"
  id="test-id"
>
  <div
    className="sc-gKAaRy eBOJRP"
  >
    <picture
      className="sc-hKFxyN eUmrnL"
    >
      <source
        srcSet="https://i.ytimg.com/vi/test-id/default.jpg"
      />
      <source
        srcSet="https://i.ytimg.com/vi/test-id/hqdefault.jpg"
      />
      <img
        className="sc-eCApnc kckRez"
        loading="lazy"
        srcSet="https://i.ytimg.com/vi/test-id/maxresdefault.jpg"
      />
    </picture>
    <div
      className="sc-iCoGMd laBOKL"
    >
      <span>
        Lorem Ipsum LIVE 2021
      </span>
      <span>
        Audio (.weba)
      </span>
    </div>
  </div>
  <div
    className="sc-dlnjwi kRnArM sc-fujyAs dSZxZQ"
  >
    <div
      className="sc-gtsrHT jBOcjR"
    />
  </div>
</div>
`;

exports[`Video renders with switch to tab 1`] = `
<div
  className="sc-jSFjdj ewagkn"
  id="test-id"
  title="Switch to tab"
>
  <div
    className="sc-gKAaRy eBOJRP"
  >
    <picture
      className="sc-hKFxyN eUmrnL"
    >
      <source
        srcSet="https://i.ytimg.com/vi/test-id/default.jpg"
      />
      <source
        srcSet="https://i.ytimg.com/vi/test-id/hqdefault.jpg"
      />
      <img
        className="sc-eCApnc kckRez"
        loading="lazy"
        srcSet="https://i.ytimg.com/vi/test-id/maxresdefault.jpg"
      />
    </picture>
    <div
      className="sc-iCoGMd laBOKL"
    >
      <span>
        Lorem Ipsum LIVE 2021
      </span>
      <span>
        Audio (.weba)
      </span>
    </div>
  </div>
</div>
`;

exports[`Video renders without error 1`] = `
<div
  className="sc-jSFjdj kNXPLX"
  id="test-id"
>
  <div
    className="sc-gKAaRy eBOJRP"
  >
    <picture
      className="sc-hKFxyN eUmrnL"
    >
      <source
        srcSet="https://i.ytimg.com/vi/test-id/default.jpg"
      />
      <source
        srcSet="https://i.ytimg.com/vi/test-id/hqdefault.jpg"
      />
      <img
        className="sc-eCApnc kckRez"
        loading="lazy"
        srcSet="https://i.ytimg.com/vi/test-id/maxresdefault.jpg"
      />
    </picture>
    <div
      className="sc-iCoGMd laBOKL"
    >
      <span>
        Lorem Ipsum LIVE 2021
      </span>
      <span>
        Audio (.weba)
      </span>
    </div>
  </div>
</div>
`;
